src/genome.py

```diff
@@ -88,154 +97,260 @@ def __str__(self) -> str:
         already contains TEs, then that TE should be disabled and
         removed from the set of active TEs.
-
         Returns a new ID for the transposable element.
         """
-        ...  # FIXME
-        return -1
+        for ft, end in self.iter_pos():
+            if end <= pos:
+                continue
+            else:
+                half_1 = ft.val.length -end + pos
+                half_2 = end - pos
+                self.insert_te_at_pos(length, ft, half_1, half_2)
+                return self.counter
+        return None
+
+    def insert_te_at_pos(self, length, ft, half_1, half_2):
+        if  ft.val.ft == 1:
+            self.disable_feature(ft)
+        insert_next(ft, Ft(1, length))
+        ft.val = Ft(ft.val.ft, half_1)
+        insert_next(ft.next, Ft(ft.val.ft, half_2))
+        self.counter += 1
+        self.active[self.counter] = ft.next
+
+    def disable_feature(self, ft):
+        ft.val = Ft(2, ft.val.length)
+        for identifier, disable in self.active.items():
+            if disable is ft:
+                self.active.pop(identifier)
+                break
+
 
     def copy_te(self, te: int, offset: int) -> int | None:
         """
         Copy a transposable element.
-
         Copy the transposable element te to an offset from its current
         location.
-
         The offset can be positive or negative; if positive the te is copied
         upwards and if negative it is copied downwards. If the offset moves
         the copy left of index 0 or right of the largest index, it should
         wrap around, since the genome is circular.
-
         If te is not active, return None (and do not copy it).
         """
-        ...  # FIXME
+
+        ft = self.active[te]
+        temp = self.active[te]
+        
+        if temp.val.ft != 1:
+            return None
+
+        if offset > 0:
+            offset = offset - temp.val.length
+            while offset > 0:
+                ft = getattr(ft, "next")
+                if ft is self.head:
+                    ft = getattr(ft, "next")
+                offset -= ft.val.length
+            half_1 = offset + ft.val.length
+            half_2 = abs(offset)
+            self.insert_te_at_pos(temp.val.length, ft, half_1, half_2)
+        else:
+            offset *= -1
+            while offset > 0:
+                ft = getattr(ft, "prev")
+                if ft is self.head:
+                    ft = getattr(ft, "prev")
+                offset -= ft.val.length
+            self.insert_te_at_pos(temp.val.length, ft, abs(offset), offset + ft.val.length)
+        return self.counter
 
     def disable_te(self, te: int) -> None:
         """
         Disable a TE.
-
         If te is an active TE, then make it inactive. Inactive
         TEs are already inactive, so there is no need to do anything
         for those.
         """
-        ...  # FIXME
+        ft = self.active.pop(te)
+        if ft is not None:
+            ft.val = Ft(2, ft.val.length)
```

Comment:
Is this necessary? Once the ID is removed from `self.active`, I would expect it to be inactive already. Then you can handle the `'x'` in the output.

And I see that you _do_ handle the `'x'` in the output, and in a fine way, but you use inactivated TEs as the default if `ft.val` is not zero ro one. There is no need to insert `Ft(2, ft.val.length)` here; and it is weird since the `ft.val` is an integer elsewhere. You can just set `ft.val = 2`.